src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -162,29 +171,23 @@ public override void Load()
 
         public async Task RefreshAsync(CancellationToken cancellationToken)
         {
+            bool success = false;
+
             // Ensure that concurrent threads do not simultaneously execute refresh operation. 
             if (Interlocked.Exchange(ref _networkOperationsInProgress, 1) == 0)
             {
                 try
                 {
-                    // Check if initial configuration load had failed
-                    if (_applicationSettings == null)
-                    {
-                        if (InitializationCacheExpires < DateTimeOffset.UtcNow)
-                        {
-                            InitializationCacheExpires = DateTimeOffset.UtcNow.Add(MinCacheExpirationInterval);
-                            await LoadAll(ignoreFailures: false, cancellationToken).ConfigureAwait(false);
-                        }
-
-                        return;
-                    }
-
-                    await RefreshIndividualKeyValues(cancellationToken).ConfigureAwait(false);
-                    await RefreshKeyValueCollections(cancellationToken).ConfigureAwait(false);
-                    await RefreshKeyValueAdapters(cancellationToken).ConfigureAwait(false);
+                    await ExecuteWithFailOverPolicyAsync((client) => RefreshAsync(client, cancellationToken), cancellationToken).ConfigureAwait(false);
+                    success = true;
                 }
                 finally
                 {
+                    if (!success)
+                    {
+                        UpdateCacheExpirationForWatchers(success: false);
+                    }

```

Comment:
If I have multiple watchers and some watchers have refreshed successfully, this code will (re)update their cache expiration as a failure case.

Reply:
If I only have one endpoint, this is a change of behavior from existing code.

Reply:
I modified the logic a little bit. Now, it will only update the cache expiration time for the watchers if all of them were successfully updated. (Partial updates won't be persisted, so cache expiration would be updated for all as a failure when some of the watchers were successful).